### PR TITLE
llama : quantize up to 31% faster on Linux with mmap

### PR DESCRIPTION
**Pull request by cebtenzzre**
This is a follow-up to #3115. It enables mmap for quantize on Linux, since no one seems to have reported a performance decrease on that platform. Windows has not been tested, and macOS has seen both a speed-up and a slow-down.

### CODE REVIEW

**Comment by bobqianic** [comment]
How does incorporating `mmap` lead to faster quantization? Can anyone explain?

**Comment by Cebtenzzre** [comment]
When the 'quantize' script reads from disk, it normally has to load a whole tensor into memory before it can start converting it to f32 and quantizing it. This change allows the input tensor to be paged in on-demand in 4096-byte chunks so it can be read and converted simultaneously.

**Comment by slaren** [comment]
I tested this with 7B f16 to q4_0. On Windows and got ~15% faster times with mmap when the model is cached, no difference when it is not cached. Under WSL2, mmap is always about ~35% faster, cached or uncached. So I think mmap can be enabled on Windows too.

File: llama.cpp [hunk 1]
```diff
@@ -5658,7 +5658,18 @@ static void llama_model_quantize_internal(const std::string & fname_inp, const s
         nthread = std::thread::hardware_concurrency();
     }
 
-    std::unique_ptr<llama_model_loader> ml(new llama_model_loader(fname_inp, /*use_mmap*/ false));
+    // mmap consistently increases speed Linux, is inconsistent on macOS
+    // (possibly related to free memory), and has not been tested on Windows.
+#ifdef __linux__
```
Comment (slaren) [inline llama.cpp:5663]:
```suggestion
#if defined(__linux__) || defined(_WIN32)
```

**Review by slaren — APPROVED** [review]

**Comment by ggerganov** [comment]
Let me run a few tests this week and we can merge.

**Review by ggerganov — APPROVED** [review]
On M1 Pro with 32GB, quantizing 13B with `mmap` enabled is ~x2 slower, so let's leave `mmap` off on Mac until we figure out something that would always improve the performance, regardless of the model size

**Pull request merged**